### DISCUSSION TITLE
Experiencing clicking noise with i2S-i2S example

### DISCUSSION THREAD

**Discussion by yujonglee** [pschatzmann/arduino-audio-tools#931] [Q&A] [answered]
I have two ESP32 connected with I2S. One is generating sinewave like [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-i2s/streams-generator-i2s.ino).

My actual code:
```cpp
#include "AudioTools.h"
#include "config.hpp"

AudioInfo info(16000, 1, 16);
SineWaveGenerator<int16_t> sineWave(32000);
GeneratedSoundStream<int16_t> sound(sineWave);
I2SStream out; 
StreamCopy copier(out, sound);

void setup(void) {  
  Serial.begin(115200);
  while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  auto config = out.defaultConfig(TX_MODE);
  config.copyFrom(info);
  
  config.port_no = I2S_NUM_1;
  config.i2s_format = I2S_STD_FORMAT;
  config.pin_ws = LISTENER_I2S_INNER_LRC;
  config.pin_bck = LISTENER_I2S_INNER_BCLK;
  config.pin_data = LISTENER_I2S_INNER_LDATA;

  out.begin(config);
  sineWave.begin(info, N_B4);
}

void loop() {
  copier.copy();
}
```

And the other is receiving sound and send it to DAC through I2S like [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-i2s-2/streams-i2s-i2s-2.ino).

My actual code:
```cpp
#include "AudioTools.h"
#include "config.hpp"

AudioInfo info(16000, 1, 16);
I2SStream in;
I2SStream out; 
StreamCopy copier(out, in);

void setup(void) {  
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning); 

  auto config_in = in.defaultConfig(RX_MODE);
  config_in.copyFrom(info); 
  config_in.i2s_format = I2S_STD_FORMAT;
  config_in.is_master = true;
  config_in.port_no = I2S_NUM_0;
  config_in.pin_bck = SPEAKER_I2S_INNER_BCLK;
  config_in.pin_ws = SPEAKER_I2S_INNER_LRC;
  config_in.pin_data = SPEAKER_I2S_INNER_LDATA;
  in.begin(config_in);

  auto config_out = out.defaultConfig(TX_MODE);
  config_out.copyFrom(info); 
  config_out.i2s_format = I2S_STD_FORMAT;
  config_out.is_master = true;
  config_out.port_no = I2S_NUM_1;
  config_out.pin_bck = SPEAKER_I2S_OUTER_BCLK;
  config_out.pin_ws = SPEAKER_I2S_OUTER_LRC;
  config_out.pin_data = SPEAKER_I2S_OUTER_SDATA;
  out.begin(config_out);
}

void loop() {
  copier.copy();
}
```

And now I got strange noise, and varying sound. Here's recorded audio with my iPhone:
https://github.com/pschatzmann/arduino-audio-tools/assets/61503739/4fc69dfd-ca19-46a3-8d21-837213146929

Notes:
1. My setup is `Web Server -> ESP1(LISTENER) -> ESP2(SPEAKER) -> DAC -> AMP/Speaker`.
2. At `22second` in recorded audio, I can hear clear sound. But rest of the audio is noise and sound keep changing
3. If I just generate sine wave in second ESP32 and directly send it to DAC, I got consistent clear sound. Problem is when I send it through `ESP32 -> ESP32 -> DAC`.
4. I am doing this with my custom board. Three gray selected line is I2S.
<img width="400" alt="Screenshot 2023-07-18 at 8 32 35 PM" src="https://github.com/pschatzmann/arduino-audio-tools/assets/61503739/15f078ae-9eff-4759-b15b-46f8c9f334ac">

**Comment by yujonglee** [comment]
After some search, I feel like this is related to `master - slave` and `clock sync`. Many people mentioned that `ESP32 I2S Slave` is not working well.([ex](https://electronics.stackexchange.com/questions/535422/can-someone-help-me-explain-i2s-timings-when-it-comes-to-master-and-multiple-sla)). 

Can anyone shed any light on this?

**Comment by yujonglee** [comment]
Update: when I set `is_master=false` in the `ESP1(Listener)`(Audio Sender), I got much better sound. But still getting weird ticking noise.

https://github.com/pschatzmann/arduino-audio-tools/assets/61503739/edd74e6f-8d50-4a68-95fc-66a5eb6200ae

**Reply by yujonglee** [reply]
Replace `SineWaveGenerator` with `SilenceGenerator`, and got only ticking sound:


https://github.com/pschatzmann/arduino-audio-tools/assets/61503739/2925467d-b861-4c86-846f-3c2f2400fca0

Starting to think maybe this is related to DMA buffer. ([reference](https://community.nxp.com/t5/Kinetis-Microcontrollers/K20-i2s-SAI-via-DMA-quot-clicking-quot-problem/m-p/404822))

**Comment by pschatzmann** [comment, marked as answer]
Connecting 2 different ESP32 might not work because of the difference in the clock speeds, but using an I2S connection on the same ESP32 might work. Did you try to set config_in.use_apll and config_out.use_apll both to the same value ? true or false

**Reply by yujonglee** [reply]
Thanks. Using `use_apll = true` seems to reduce the noise, but not completely.

> Connecting 2 different ESP32 might not work because of the difference in the clock speeds.

To solve this, should I get external clock and feed it as master clock for both ESP32?
